---
title: "Microbiome Data Analysis"
author: "Kaiden Liu"
date: "26/05/2021"
output:
  beamer_presentation:
    colortheme: "dolphin"
    fonttheme: "structurebold"
---

Phyloseq is good for storing complex phylogenetic sequencing data 

&nbsp;

Operational Taxonomic Unit (OTU): groups of closely related individuals

# 1) OTU table  

\scriptsize
* row: taxa/OTU
    + groups of closely related individuals
* column: sample
* value: the number of reads

\normalsize

\tiny
```{r}
set.seed(526)
otumat = matrix(sample(0:100, 100, replace = TRUE,prob=c(0.5,rep(0.5/100,100))), nrow = 10, ncol = 10)
rownames(otumat) <- paste0("OTU", 1:nrow(otumat))
colnames(otumat) <- paste0("Sample", 1:ncol(otumat))
otumat
```
\normalsize

***

# 2) taxonomy table   

Table of names of the taxonomic rank of the data

* row: OTU/taxonomy  
* column: taxonomy rank(levels)  
* value: names of the taxonomy (family)  

&nbsp;

\scriptsize 
```{r}
taxmat = matrix(sample(letters, 70, replace = TRUE), 
                nrow = nrow(otumat), ncol = 7)
rownames(taxmat) <- rownames(otumat)
colnames(taxmat) <- c("Domain", "Phylum", "Class", 
                      "Order", "Family", "Genus", "Species")
taxmat
```
\normalsize

***

# 2.5) Creating a phyloseq object

\scriptsize
```{r}
library("phyloseq")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)

physeq = phyloseq(OTU, TAX)
physeq
```
\normalsize

***

# 3) sample variables  

* Location: Location of where the sample is collected(eg: Feces, Blood, Skin)
* Depth: Number of sample sequenced

\scriptsize
```{R}
set.seed(999)
sampledata = sample_data(data.frame(
  Location = sample(LETTERS[1:4], size=nsamples(physeq), replace=TRUE),
  Depth = sample(50:1000, size=nsamples(physeq), replace=TRUE),
  row.names=sample_names(physeq),
  stringsAsFactors=FALSE
))
sampledata
```
\normalsize
***

## 4) phylogenetic Tree

Shows how the different taxa are related.  

\scriptsize
```{R}
library("ape")
random_tree = rtree(ntaxa(physeq), rooted=TRUE, 
                    tip.label=taxa_names(physeq))
plot(random_tree)
```
\normalsize

***

# 4.5) Complete the phyloseq object by merging the two new "tables"

\scriptsize
```{R}
physeq1 = merge_phyloseq(physeq, sampledata, random_tree)
physeq1
```
\normalsize

***

We can also display the new phylogenetic tree with our new phyloseq object

\footnotesize
```{r echo=F}
plot_tree(physeq1, color="Depth", shape="Location", label.tips="taxa_names", ladderize="right", plot.margin=0.3)
```
\normalsize

***
\footnotesize
Value is the measure of support of the node, calculated by bootstrapping

1) We draw samples from the data with replacement for a specified size
2) we train a model with the samples, and fit the model to the data again.
3) calculate the "accuracy" of the result

```{r, echo=F}
rich_dense_biom  = system.file("extdata", "rich_dense_otu_table.biom",  package="phyloseq")
treefilename = system.file("extdata", "biom-tree.phy",  package="phyloseq")
refseqfilename = system.file("extdata", "biom-refseq.fasta",  package="phyloseq")

myData = import_biom(rich_dense_biom, treefilename, refseqfilename, parseFunction=parse_taxonomy_greengenes)

plot_tree(myData, color="Genus", shape="BODY_SITE", size="abundance")
```
\normalsize

***

```{r, warning=F}
plot_heatmap(physeq1)
```

&nbsp; 

# 5) Reference Seq  

This table would give us more details about our data

```{R}
refseq(myData)
```

&nbsp;

ex) DNAStringSet, RNAStringSet, and AAStringSet from Biostrings package

****

# Workflow for Microbiome Data Analysis

FASTQ format file contains biological sequence and the corresponding quality score.

With DADA2, we want to convert this FASTQ file to an OTU table.  

* We want to replace OTUs with ASVs(Amplicon sequence variant)
    + Higher resolution
    + Higher Accuracy
    + Linear Computation Time
    
&nbsp;

But before we convert...

***

# Trim

plotQualityProfile:

- inspect fastq file quality  
- underlying heatmap shows frequency of each score at each position  
- green --> mean  
- orange --> quantile(dash is 25th quantile and 75th quantile)  

***

\tiny
```{r, echo=F, message=F}
set.seed(100)


#remember set working directory to ~/week 3
#setwd("~/Summer Research/week/week 3")

#remember to ask if I can use this instead
setwd("~/Summer_Research")
miseq_path <- file.path("data", "MiSeq_SOP")
filt_path <- file.path("data", "filtered")

if(!file_test("-d", miseq_path)) {
  dir.create(miseq_path)
  download.file("http://www.mothur.org/MiSeqDevelopmentData/StabilityNoMetaG.tar",
                destfile = file.path(miseq_path, "StabilityNoMetaG.tar"))
  system(paste0("tar -xvf ", file.path(miseq_path, "StabilityNoMetaG.tar"),
                " -C ", miseq_path, "/"))
}

fns <- sort(list.files(miseq_path, full.names = TRUE))
# seperate data files to R1 and R2 names
fnFs <- fns[grepl("R1", fns)]
fnRs <- fns[grepl("R2", fns)]

# Trim and Filter

#add myself
library("dada2")
library("ggplot2")

print(plotQualityProfile(fnFs[1]) + ggtitle("Fwd"))


```
\normalsize



Idea: We want to truncate the read based on this plot so that the quality scores stay near the top
    
***

# Data Manipulation

## Taxonomic filtering

- We want to remove the data that are rare in our taxonomy table, because they are not likely to be true in nature

## Prevalence Filtering

prevalence: fraction of total samples in which a taxa is observed

- Identify and filter outlier

## Agglomerating taxa

When the species are categorized too deep and starting to be redundant, we want to group the data back together by how closely related they are in terms of taxa.

- Figure 4

## Abundance value transformation

The challenge of different library sizes among the samples can be accounted by transforming the count data to proportions or relative abundances.

- Figure 5

***

# Ordination plots

## principal coordinates analysis (PCoA)

We want to map our data from a high dimension to a low dimension, so we can visualize the similarities of data(by how close they are).

 - First axis > Second Axis (variability)

Figure 10

***

## Distance matrix

We can summarize the relationship between points with distance

1) Bray-Curtis dissimilarity

- based on counts 
- ranges from 0 to 1
    + 0 means the two samples are from the same group
    + 1 means the two samples are different
- not a distance

2) Weighted UniFrac Distance

- based on the phylogenetic distance
- edges of the phylogenetic tree are weighted proportional to the abundance of the taxa
- is a distance

*** 

## PCA on ranks

- represent  abundances by ranks
    + taxa with smallest in sample maps to 1, second smallest sample maps to 2
    
- Good for data with heavy-tailed

- Threshold for small abundance (absent data) --> large difference in rank

Figure 15

***

# Network analysis

Create a network by thresholding a distance matrix

&nbsp;

## Minimum spanning tree

We assign weights to all the edges of the network. 

We want to find a way to connect all dots together without any cycles, and with the smallest sum of edge weights.

- Nearest neighbors

&nbsp;

Picture: https://en.wikipedia.org/wiki/Minimum_spanning_tree#/media/File:Minimum_spanning_tree.svg

&nbsp;

Figure 23  

&nbsp;

pure edges: edges that connects two nodes of the same level

*** 

### Graph-based two-sample tests

Null hypothesis: two samples come from the same distribution

&nbsp;

test statistics: number of pure edges

&nbsp;

histogram: permute sample type randomly and construct a histogram

- if number of pure edges is more than the test statistics, we reject the null hypothesis.


***

# Supervised learning

1) Partial Least Square

2) Random Forest

***
\tiny
```{r, message=F}
setup_example<- (c("phyloseq", "ggplot2", "plyr", "dplyr", "reshape2",
                "ade4", "ggrepel", 'randomForest', 'testingfail'))
lapply(setup_example,require,character.only=T)
```
\normalsize